### ISSUE TITLE
[HOLD for payment 2023-04-14] [$1000] There is no feature to remove preferred pronouns once a user adds preferred pronouns

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16349]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Click on profile avatar
2. Click on profile
3. Click on Pronouns
4. Search pronoun
5. Select a pronoun from the search results list
6. Try to remove the applied pronoun

## Expected Result:
 A feature to remove the added pronoun should be available

## Actual Result:
There is no feature that supports removing applied pronouns, the available option is changing pronouns only

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.87-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/226667129-db68a658-7e8a-416b-bbd9-225be1bb84d5.mp4

https://user-images.githubusercontent.com/43996225/226667184-c8c41518-69aa-472f-81e8-8a647421de80.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Natnael-Guchima
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1679123540639329

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0179e823ce97e8d1de</li>
        <li>Upwork Job ID: 1638619807394324480</li>
        <li>Last Price Increase: 2023-03-22</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by varshamb** [comment]
## Proposal
New feature to unselect the Pronouns.
### What is the root cause of that problem?
There is no way to unselect an option.
### What changes do you think we should make in order to solve the problem?
We need to check if new selected value of Pronouns is same as current value, if it is same then pass NULL value to API.

**Comment by MelvinBot** [comment]
📣 @varshamb! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by Natnael-Guchima** [comment]
Contributor details
Your Expensify account email: natnaelguchima12gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0135b3a84ec2709c21

**Comment by MelvinBot** [comment]
⚠️ Missing/invalid email or upwork profile link. Please make sure you add both your Expensify email and Upwork profile link in the format specified.

**Comment by varshamb** [comment]
Contributor details
Your Expensify account email: varshamadankar@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~010aadecc66d4e21a1

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by trjExpensify** [comment]
@puneetlath you happy with this proposed pattern for deselect? It seems pretty straightforward to me.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0179e823ce97e8d1de

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @mollfpr (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcochavezf (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by dhairyasenjaliya** [comment]
### Proposal

**Please re-state the problem that we are trying to solve in this issue.**
 
 - There is no feature to remove preferred pronouns once a user adds preferred pronouns 
 
**What is the root cause of that problem?**
 
 - After refactoring the Pronouns component we never add any validation or any feature to remove the selected Pronouns value 
 
 - We are also not displaying the current selected Pronouns to the user 
 
 - Currently, we are using `<OptionsSelector>` for search input adding a default name and the clear button will definitely look weird 

**What changes do you think we should make in order to solve the problem?**
 
 - IMO we should re-use the `<TestToolRow>` component to display the Pronouns name and clear button which will look like the one below that actually blends well with the current design 
 
<img width="373" alt="Screenshot 2023-03-23 at 9 50 51 AM" src="https://user-images.githubusercontent.com/47522946/227102268-96b4c549-a510-462a-bf38-5d74433a20aa.png">

- We are already using  `<TestToolRow>` component for invalid auth status and destroy credential 

<img width="292" alt="Screenshot 2023-03-23 at 10 08 19 AM" src="https://user-images.githubusercontent.com/47522946/227104671-52adfbe6-f5ff-4fd3-9d1c-a024e7c3ffb8.png">


- In order to show the currently selected pronouns by the user we should define one function that returns selected pronouns like below and add it to the `PronounsPage.js`

```diff  
PronounsPage.js
+    /**
+     * Fetch the actually selected pronouns from currentUserPersonalDetails
+     *
+     * @returns {string}
+    */
+    getPronouns = () => {
+       let pronounsKey = lodashGet(this.props.currentUserPersonalDetails, 'pronouns', '');
+        if (pronounsKey.startsWith(CONST.PRONOUNS.PREFIX)) {
+            pronounsKey = pronounsKey.slice(CONST.PRONOUNS.PREFIX.length);
+       }
+        return lodashGet(this.props.translate('pronouns'), pronounsKey, null);
+    };

```

- Once the `getPronouns()` is defined we need to add `<TestToolRow>` we will render this component if there are any selected pronouns founded 

```diff
+ const selectedPronouns = this.getPronouns();
+ <TestToolRow title={selectedPronouns}>
+      <Button
+           small
+           text="Clear Pronouns"
+           onPress={() => this.updatePronouns('')} // Clear the selected Pronouns
+      />
+ </TestToolRow>
```

- Note other changes can be done at PR 

**Result**

https://user-images.githubusercontent.com/47522946/227103958-cd5fd992-5a91-45f2-b112-fdae7ca0cba5.mp4


cc @marcochavezf

**Comment by MelvinBot** [comment]
Auto-assign attempt failed, all eligible assignees are OOO.

**Comment by puneetlath** [comment]
Ah looks like Shawn is OOO, but @shawnborton when you're back perhaps you can give us some thoughts on what you think the best approach is for this given that there is no existing pattern in the app.

**Comment by puneetlath** [comment]
That makes sense to me. @trjExpensify feel free to change if you disagree.

**Comment by trjExpensify** [comment]
Yeah, I'm not sure I agree with the clear pronouns button, it seems pretty heavy. Whatever we do here should extend to other places we'll be using the list pattern i.e categories, tags etc. 

I think tapping the selected value in the list to deselect it works, IMO. It's kinda' subtle, but it would make sense that when you tap the value that has a green tick, it deselects that value.

**Comment by dhairyasenjaliya** [comment]
Hey @trjExpensify would you able to check this suggestion https://github.com/Expensify/App/issues/16349?notification_referrer_id=NT_kwDOAtUkgrM1OTI3NzgxNDk1OjQ3NTIyOTQ2&notifications_query=reason%3Aparticipating#issuecomment-1480594170

- Seems if we add tap to de-select value its bit weird behavior since a user manually need to search selected value and then un-select in most case user won't know that they can un-select value this way , so my suggestion was simple to display clear button at top if pronouns selected

**Comment by situchan** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
No feature to remove preferred pronoun user selected

### What is the root cause of that problem?
We missed this case when we implement (or refactor) Pronouns page

### What changes do you think we should make in order to solve the problem?
- Similar to Timezone page, auto-populate user's current preferred pronoun on search input when enter Pronouns page.
- After selecting the item with green checkmark (which is user's current pronoun), clear it.

Implementation detail:
- define local variable (class member) something like `this.initiallyFocusedOption = {};` in constructor
- When load pronouns, update this variable to have selected value and key if `isCurrentPronouns` = true
Something like this:
```
            if (isCurrentPronouns) {
                this.initiallyFocusedOption = {
                    text: value,
                    keyForList: key,
                };
            }
```
- update initial state.searchValue in constructor to `this.initiallyFocusedOption.text || ''`
So far, pre-populate is done.
- pass `initialyFocusedOptionKey={this.initiallyFocusedOption.keyForList}` to OptionsSelector to auto-scroll selected value
This is to prevent user scroll manually to the selected pronoun when filtered result is too large to view all in screen viewport.
- bind `updatePronouns` function to access `this`
- finally call `PersonalDetails.updatePronouns` with proper value:
```
    updatePronouns(selectedPronouns) {
        PersonalDetails.updatePronouns(selectedPronouns.keyForList === this.initiallyFocusedOption.keyForList ? '' : selectedPronouns.value);
    }
```

Demo:

https://user-images.githubusercontent.com/108292595/227793804-fbfb66d8-5117-4589-9e6e-227e93725e29.mp4

**Comment by marcochavezf** [comment]
> Hey @trjExpensify would you able to check this suggestion https://github.com/Expensify/App/issues/16349?notification_referrer_id=NT_kwDOAtUkgrM1OTI3NzgxNDk1OjQ3NTIyOTQ2&notifications_query=reason%3Aparticipating#issuecomment-1480594170
> 
> * Seems if we add tap to de-select value its bit weird behavior since a user manually need to search selected value and then un-select in most case user won't know that they can un-select value this way , so my suggestion was simple to display clear button at top if pronouns selected

Still on hold for design. The clear button seems to be a good strategy for me.

**Comment by trjExpensify** [comment]
Yes, I think we should do these two things personally: 

>Similar to Timezone page, auto-populate user's current preferred pronoun on search input when enter Pronouns page.
>After selecting the item with green checkmark (which is user's current pronoun), clear it.

**Comment by dhairyasenjaliya** [comment]
@shawnborton for demo i have added clear button that we already have used for destroy credentials in preferences, I do think this would be good options here as well since we will display selected pronounce at top along with clear button that will also support current flow of pronounce optionList within addition changes 

https://github.com/Expensify/App/issues/16349#issuecomment-1480594170

**Comment by shawnborton** [comment]
I don't think the clear button makes sense for this kind of component. It looks like you are suggesting that we show the currently selected value with a clear button above the search input which can then show more values. That seems like a very heavy solution to me.

**Comment by situchan** [comment]
I am not sure it's intentional but when no search value, shows nothing. It's inconsistent from Timezone page which shows all timezones when nothing searched.

**Comment by dhairyasenjaliya** [comment]
Hm but even showing default value at first user won't be able to notice if then can actually clear values by tapping again (might be bit confusing IMO) for timezone its different case where by there will be always one value selected @shawnborton

**Comment by shawnborton** [comment]
Yup, that one works for me.

**Comment by mollfpr** [comment]
Thanks, @shawnborton!

Thank you for the proposal @varshamb! We should hold the proposal before its triage, and on our final decision, your proposal is still incomplete to solve this issue.

@marcochavezf @trjExpensify Let's assign @situchan to solve this issue with their [proposal](https://github.com/Expensify/App/issues/16349#issuecomment-1484167800).

🎀 👀 🎀 C+ reviewed!

**Comment by varshamb** [comment]
@mollfpr 

> According to guideline, your proposal will not be considered since it's before `Help Wanted`.

Thanks for the reply. This is a recent change to contributing.md. It was not communicated on slack.

I wasn't aware of it, so please consider the proposal to be fair with my time.

cc: @tjferriss @marcochavezf

**Comment by varshamb** [comment]
@mollfpr I believe expensify values our time. We shouldn't be expected to fix unrelated bugs which are already reported in the same issue.



@tjferriss the accepted proposal is my proposal + a different bug solved. 

Please help us out here

**Comment by situchan** [comment]
@varshamb 
Similar case: https://github.com/Expensify/App/issues/16053#issuecomment-1479761694, followed by engineer's comment.

**Comment by situchan** [comment]
@varshamb Here's slack convo: https://expensify.slack.com/archives/C01GTK53T8Q/p1679500278205259
This was posted even before this issue was marked as `External`

Why the issues need to be inconsistent at the same day? one stick to guideline, another one not stick to guideline

Let's wait for team's decision.

**Comment by trjExpensify** [comment]
It's expected that there aren't any results until you start typing in the first instance, which the [linked slack thread](https://expensify.slack.com/archives/C049HHMV9SM/p1677816930988059?thread_ts=1677816930.988059&cid=C049HHMV9SM) discussed and deemed not to be a bug, right? 

What we're suggesting here is that when you have made a selection and come back to the Pronouns page, that selection is visible to deselect it.

**Comment by varshamb** [comment]
> What we're suggesting here is that when you have made a selection and come back to the Pronouns page, that selection is visible to deselect it.

@trjExpensify that's a different bug. I reported it 4 days ago https://expensify.slack.com/archives/C049HHMV9SM/p1679631777231259

**Comment by mollfpr** [comment]
Sorry @varshamb, we updated the CONTRIBUTING.md weeks ago https://github.com/Expensify/App/pull/15780/files#diff-f50bedf30868fb94f5360486528e898c02e4cbceb82cdd72b9057c2b113d7b84R114 and https://github.com/Expensify/App/issues/16053#issuecomment-1479761694 this case is already clear what step we took.

Just to let you know, there's no guarantee that your time will be compensated.

>Please be aware that compensation for any support in solving an issue is provided **entirely at Expensify’s discretion**. Personal time or resources applied towards investigating a proposal **will not guarantee compensation**. Compensation is only guaranteed to those who **[propose a solution and get hired for that job](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job)**. 

>We understand there may be cases where a selected proposal may take inspiration from a previous proposal. Unfortunately, it’s not possible for us to evaluate every individual case and we have no process that can efficiently do so. Issues with higher rewards come with higher risk factors so try to keep things civil and make the best proposal you can. Once again, **any information provided may not necessarily lead to you getting hired for that issue or compensated in any way.**

**Comment by trjExpensify** [comment]
We can take the design hold off this issue now.

**Comment by trjExpensify** [comment]
@situchan is there a PR for this yet?

**Comment by situchan** [comment]
> @situchan is there a PR for this yet?

I am not assigned yet. Waiting for @marcochavezf's approval

**Comment by marcochavezf** [comment]
Hi guys, sorry for the delay here. Assigning @situchan

**Comment by marcochavezf** [comment]
> I wasn't aware of it, so please consider the proposal to be fair with my time.

@varshamb thank you for bringing this to our attention. We understand that everyone's time is valuable, and we want to ensure that our system takes that into consideration. We're planning to expand our system to consider the effort a contributor spends to figure out a solution (even if the proposal is not selected). Meanwhile, we encourage you to keep posting proposals on other reported bugs and sharing your thoughts with us.

**Comment by MelvinBot** [comment]
📣 @situchan You have been assigned to this job by @marcochavezf!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by situchan** [comment]
@mollfpr @marcochavezf [PR](https://github.com/Expensify/App/pull/16872) is ready for review

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.96-4 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16872


If no regressions arise, payment will be issued on 2023-04-14. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mollfpr / @marcochavezf] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mollfpr / @marcochavezf] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mollfpr / @marcochavezf] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@trjExpensify] Determine if we should create a regression test for this bug.
- [ ] [@mollfpr] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by marcochavezf** [comment]
We didn't have the option to unselect the preferred pronouns since the feature was implemented, so technically, no bug introduced this behavior. Also, I think the regression test steps will be enough to prevent this issue in the future. I checked off the corresponding checkboxes.

**Comment by trjExpensify** [comment]
Settled up with @situchan 👍

**Comment by mollfpr** [comment]
@trjExpensify Accepted, thank you!

**Comment by Natnael-Guchima** [comment]
@trjExpensify Accepted. Thank you!

**Comment by trjExpensify** [comment]
Cool, settled up with everyone. Thanks!

**Comment by Beamanator** [comment]
> > [@trjExpensify] Determine if we should create a regression test for this bug.
> 
> @zanyrenney @twisterdotcom @Beamanator presumably, once the account settings project is complete, the regression tests will be added to accommodate the changes right? Given that, I don't think I'll add to a competing standalone one here for removing pronouns.

Agreed @trjExpensify - that's one of the final tasks in the project, hoping that will happen soon - making sure all the TestRail steps are available for all the new pages 🙏  We've been adding a few as we go, but will definitely do a final sweep at the end 👍

**Comment by trjExpensify** [comment]
Awesome, that's great!
